rts/motoko-rts/src/gc/mark_compact/bitmap.rs

```diff
@@ -4,37 +4,45 @@ use crate::types::{size_of, Blob, Bytes, Obj};
 
 /// Current bitmap
 static mut BITMAP_PTR: *mut u8 = core::ptr::null_mut();
+static mut BITMAP_COMPENSATION: usize = 0;
+static mut BITMAP_SIZE: u32 = 0;
 
-pub unsafe fn alloc_bitmap<M: Memory>(mem: &mut M, heap_size: Bytes<u32>) {
+pub unsafe fn alloc_bitmap<M: Memory>(mem: &mut M, heap_size: Bytes<u32>, heap_prefix_words: u32) {
+    debug_assert_eq!(heap_prefix_words % 8, 0);
     // We will have at most this many objects in the heap, each requiring a bit
     let n_bits = heap_size.to_words().as_u32();
     // Each byte will hold 8 bits.
-    let bitmap_bytes = (n_bits + 7) / 8;
+    BITMAP_SIZE = (n_bits + 7) / 8;
     // Also round allocation up to 8-bytes to make iteration efficient. We want to be able to read
     // 64 bits in a single read and check as many bits as possible with a single `word != 0`.
-    let bitmap_bytes = Bytes(((bitmap_bytes + 7) / 8) * 8);
+    let bitmap_bytes = Bytes(((BITMAP_SIZE + 7) / 8) * 8);
     // Allocating an actual object here as otherwise dump_heap gets confused
     let blob = alloc_blob(mem, bitmap_bytes).get_ptr() as *mut Blob;
     memzero(blob.payload_addr() as usize, bitmap_bytes.to_words());
 
-    BITMAP_PTR = blob.payload_addr()
+    BITMAP_COMPENSATION = (heap_prefix_words / 8) as usize;
+    BITMAP_PTR = blob.payload_addr().sub(BITMAP_COMPENSATION)
 }
 
 pub unsafe fn free_bitmap() {
     BITMAP_PTR = core::ptr::null_mut();
+    BITMAP_SIZE = 0;
+    BITMAP_COMPENSATION = 0
 }
 
 pub unsafe fn get_bit(idx: u32) -> bool {
-    let byte_idx = idx / 8;
+    let (byte_idx, bit_idx) = (idx / 8, idx % 8);
+    debug_assert!(byte_idx as usize >= BITMAP_COMPENSATION);
+    debug_assert!(BITMAP_COMPENSATION + BITMAP_SIZE as usize > byte_idx as usize);
     let byte = *BITMAP_PTR.add(byte_idx as usize);
-    let bit_idx = idx % 8;
-    (byte >> bit_idx) & 0b1 == 0b1
+    (byte >> bit_idx) & 0b1 == 0b1 // TODO IDEA: != 0 (faster, also below!)
```

Comment:
This is a trivial change, why leave a TODO instead of just doing it?

Why do you think `(byte >> bit_idx) & 0b1 != 0` is faster that `(byte >> bit_idx) & 0b1 == 0b1`? I'd expect the compiler to do this transformation if it's obviously faster.

What does "also below!" mean?

Reply:
I believe LLVM would easily optimise this. f1b9fb7, because it is shorter.